### Task 1
A programmer creates a remote-controlled robot and wants to find out how many steps it
takes to exit a maze.
The maze is represented by a 6 by 6 square grid. Each position in the grid is represented by
a pair of coordinates. The top left square display has x = 0 and y = 0.
The robot moves left, right, up or down according to a direction entered. The following are
valid inputs:

- "U"  : up
- "D"  : down
- "L"  : left
- "R"  : right
- ' ' (empty string)  : continue with prev move, else do nothing

When a direction is entered, the robot moves one position in that direction. After the robot
moves, the position it was previously on is replaced by a 'X'. The robot cannot move to the
same spot twice. If the direction would place the robot on a wall or a position previously
stepped on, the robot does not move. The maze is displayed after each move.

The robot is denoted by 'T', the walls '#' and empty space '.'.

    # # # # T #
    # . # . . #
    # . . . # #
    # # . # . #
    # . . . . #
    # # # # # # 

#### Task 1.1
Using the maze given in MAZE.txt, write program code to:

- read the maze from the text file and store it in a suitable array structure
- randomize the exit along the last row of the maze
- update the exit square on the grid with a '.'
- display the maze when the robot is in its initial position at x = 4 and y = 0.

Test the program and show the output.  [6]

In [67]:
# Task 1.1
# Started 12:24PM

import random
FILENAME = "MAZE.txt"

def print_maze(maze):
    for row in maze:
        print(" ".join(row))

# Load maze from file
maze = [["" for _ in range(6)] for _ in range(6)]
with open(FILENAME, "r") as file:
    # File will auto-close
    y = 0
    for line in file:
        maze[y] = list(line.strip())
        y += 1

# Randomize exit
exit = random.randint(1, 4)

# Place exit in maze
maze[-1][exit] = "."

x, y = 4, 0
maze[y][x] = "T"

print_maze(maze)

# # # # T #
# . # . . #
# . . . # #
# # . # . #
# . . . . #
# # . # # #


#### Task 1.2
Add to your program code to:

- take in and validate a direction
- calculate a new position
- check if this position is an empty space ('.')
- update the grid so that the previous position of 'T' is replaced with a 'X' and the robotis located in its new position
- display the maze
- continue this until the robot is moved to the exit of the maze
- when robot is at the exit, the number of steps taken is displayed.

Test run the program.

In [68]:
# Task 1.2

print_maze(maze)

running = True
previousx, previousy = 0, 0
while running:
    instr = input("Move> ")
    print()
    dx, dy = 0, 0
    valid = True
    match instr:
        case "L":
            dx = -1
        case "R":
            dx = 1
        case "U":
            dy = -1
        case "D":
            dy = 1
        case "":
            dx, dy = previousx, previousy
        case _:
            print("Invalid move")
            valid = False
            
    if valid:
        previousx, previousy = dx, dy
        # nx, ny: new coordinates
        nx, ny = x+dx, y+dy
        
        if ny < 0:
            print("Reached edge of map")
        elif ny == 5:
            running = False
            maze[ny][nx] = "T"
            maze[y][x] = "X"
        elif maze[ny][nx] != ".":
            print("Tried walking into a wall")
        else:
            maze[ny][nx] = "T"
            maze[y][x] = "X"
            x, y = nx, ny
    print_maze(maze)

# # # # T #
# . # . . #
# . . . # #
# # . # . #
# . . . . #
# # . # # #


Move>  D



# # # # X #
# . # . T #
# . . . # #
# # . # . #
# . . . . #
# # . # # #


Move>  L



# # # # X #
# . # T X #
# . . . # #
# # . # . #
# . . . . #
# # . # # #


Move>  D



# # # # X #
# . # X X #
# . . T # #
# # . # . #
# . . . . #
# # . # # #


Move>  L



# # # # X #
# . # X X #
# . T X # #
# # . # . #
# . . . . #
# # . # # #


Move>  D



# # # # X #
# . # X X #
# . X X # #
# # T # . #
# . . . . #
# # . # # #


Move>  



# # # # X #
# . # X X #
# . X X # #
# # X # . #
# . T . . #
# # . # # #


Move>  



# # # # X #
# . # X X #
# . X X # #
# # X # . #
# . X . . #
# # T # # #


### Task 2
This task is to perform sorting algorithms on Person objects held in a 1-dimensional array.


#### Task 2.1
The class `Person` contains two properties:

- `name` - stored as a string
- `age` - stored as an integer

Write program code to declare the class `Person` and its constructor and `print()` method
to output the `name` and `age` of a `Person` object.  [2]

In [2]:
# Task 2.1

class Person():
    def __init__(self, name, age):
        self._name = name
        self._age = age
    def print(self):
        print(f"(Name: {self._name}, age: {self._age})")
    def __str__(self):
        return f"(Name: {self._name}, age: {self._age})"
    def get_name(self):
        return self._name
    def get_age(self):
        return self._age
    def set_name(self, name):
        self._name = name
    def set_age(self, age):
        self._age = age

#### Task 2.2
Write a function `task2_2(filename)` that:

- takes a string `filename` which represents the name of a text file
- reads in the contents of the text file
- returns the content as a list of `Person` objects.

Call the function `task2_2` with the file `PERSON.txt` and print `Person` objects using the following statements:

    list_of_person = task2_2('PERSON.txt')
    for person in list_of_person:
    person.print() 
                                                                                    [4]

In [3]:
def task2_2(filename):
    persons = []
    with open(filename, "r") as file:
        # File closes automatically
        for line in file:
            param = line.split(",")
            person = Person(param[0].strip(), param[1].strip())
            persons += [person]
    return persons

list_of_person = task2_2('PERSON.txt')
for person in list_of_person:
    person.print()

(Name: Alice, age: 18)
(Name: Bob, age: 20)
(Name: Charlie, age: 17)
(Name: David, age: 16)
(Name: Emily, age: 19)
(Name: Austin, age: 19)
(Name: Cole, age: 20)
(Name: Adam, age: 16)
(Name: Benjamin, age: 16)
(Name: Chloe, age: 19)
(Name: Daniel, age: 19)
(Name: Eva, age: 20)
(Name: Bailey, age: 18)
(Name: Daisy, age: 18)
(Name: Amelia, age: 17)
(Name: Brian, age: 19)
(Name: Catherine, age: 18)
(Name: Dylan, age: 17)
(Name: Eleanor, age: 16)
(Name: Bella, age: 17)
(Name: Caleb, age: 16)
(Name: Delilah, age: 20)
(Name: Ethan, age: 17)
(Name: Ella, age: 18)
(Name: Arthur, age: 20)


#### Task 2.3
One method of sorting is the insertion sort.
Write a function `task2_3(list_of_person, key, order)` that:

- accepts three parameters:
 - `list_of_person` contains a list of `Person` objects
 - `key` should be one of the values:
   - `name` – list to be sorted by name
   - `age` – list to be sorted by age
 - `order` should be one of the values:
   - `asc` – list to be sorted by key in ascending order
   - `desc` – list to be sorted by key in descending order
- sorts `list_of_person` by `key` in `order` using insertion sort.

Call the function `task2_3` with the contents of the file `PERSON.txt` and print the sorted`Person` objects using the following statements:

    list_of_person = task2_2('PERSON.txt' )
    task2_3(list_of_person, 'name', 'asc')
    for person in list_of_person:
        person.print()
                                                                                            [8]

In [4]:
# Task 2.3

def is_after(A, B, key, order):
    if key == "name":
        aparam, bparam = A.get_name(), B.get_name()
    elif key == "age":
        aparam, bparam = int(A.get_age()), int(B.get_age())
    compare = aparam < bparam
    asc = order=="asc"
    if asc:
        return compare
    else:
        return not compare

def task2_3(list_of_person, key, order):
    for i in range(len(list_of_person)):
        j = i
        while j != 0 and is_after(list_of_person[j-1], list_of_person[j], key, order):
            list_of_person[j], list_of_person[j-1] = list_of_person[j-1], list_of_person[j]
            j -= 1

list_of_person = task2_2('PERSON.txt')
task2_3(list_of_person, 'name', 'asc')
for person in list_of_person:
    person.print()

(Name: Adam, age: 16)
(Name: Alice, age: 18)
(Name: Amelia, age: 17)
(Name: Arthur, age: 20)
(Name: Austin, age: 19)
(Name: Bailey, age: 18)
(Name: Bella, age: 17)
(Name: Benjamin, age: 16)
(Name: Bob, age: 20)
(Name: Brian, age: 19)
(Name: Caleb, age: 16)
(Name: Catherine, age: 18)
(Name: Charlie, age: 17)
(Name: Chloe, age: 19)
(Name: Cole, age: 20)
(Name: Daisy, age: 18)
(Name: Daniel, age: 19)
(Name: David, age: 16)
(Name: Delilah, age: 20)
(Name: Dylan, age: 17)
(Name: Eleanor, age: 16)
(Name: Ella, age: 18)
(Name: Emily, age: 19)
(Name: Ethan, age: 17)
(Name: Eva, age: 20)


#### Task 2.4
Another method of sorting is the quick sort.
Write a function `task2_4(list_of_person, key, order)` that:
- accepts three parameters:
 - `list_of_person` contains a list of `Person` objects
 - `key` should be one of the values:
   - `name` – list to be sorted by name
   - `age` – list to be sorted by age
 - `order` should be one of the values :
   - `asc` – list to be sorted by key in ascending order
   - `desc` – list to be sorted by key in descending order
- sorts `list_of_person` by `key` in `order` using quick sort. 

Call the function `task2_4` with the contents of the file `PERSON.txt` and print the sorted `Person` objects using the following statements:

    list_of_person = task2_2('PERSON.txt')
    task2_4(list_of_person, 'age', 'desc')
    for person in list_of_person:
        person.print() 
                                                                                                        [8]

In [23]:
def partition(list_of_person, key, order, low, high):
    pivot = low
    pivot_value = list_of_person[pivot]

    left, right = low+1, high
    counter = 0
    while right >= left and counter < 100:
        counter += 1
        while right >= left and left < high and not is_after(list_of_person[left], pivot_value, key, order):
            left += 1
        while right >= left and right > low and is_after(list_of_person[right], pivot_value, key, order):
            right -= 1
        if right > left:
            list_of_person[right], list_of_person[left] = list_of_person[left], list_of_person[right]
    list_of_person[right], list_of_person[pivot] = list_of_person[pivot], list_of_person[right]
    return right

def task2_4(list_of_person, key, order, low=0, high=None):
    if high == None:
        high = len(list_of_person) - 1
    if low >= high:
        return
        
    pivot = partition(list_of_person, key, order, low, high)
    task2_4(list_of_person, key, order, low, pivot-1)
    task2_4(list_of_person, key, order, pivot+1, high)

list_of_person = task2_2('PERSON.txt')
task2_4(list_of_person, 'name', 'asc')
for person in list_of_person:
    print(person)

(Name: Adam, age: 16)
(Name: Alice, age: 18)
(Name: Amelia, age: 17)
(Name: Arthur, age: 20)
(Name: Austin, age: 19)
(Name: Bailey, age: 18)
(Name: Bella, age: 17)
(Name: Benjamin, age: 16)
(Name: Bob, age: 20)
(Name: Brian, age: 19)
(Name: Caleb, age: 16)
(Name: Catherine, age: 18)
(Name: Charlie, age: 17)
(Name: Chloe, age: 19)
(Name: Cole, age: 20)
(Name: Daisy, age: 18)
(Name: Daniel, age: 19)
(Name: David, age: 16)
(Name: Delilah, age: 20)
(Name: Dylan, age: 17)
(Name: Eleanor, age: 16)
(Name: Ella, age: 18)
(Name: Emily, age: 19)
(Name: Ethan, age: 17)
(Name: Eva, age: 20)


#### Task 2.5
Write a function `task2_5(list_of_person, method, key, order)` that:

- accepts four parameters:
 - `list_of_person` contains a list of `Person` objects
 - `method` should be one of the values:
   - `insertion sort` – sort the list using insertion sort
   - `quick sort` – sort the list using quick sort
 - `key` should be one of the values:
   - `name` – list to be sorted by name
   - `age` – list to be sorted by age
 - `order` should be one of the values:
   - `asc` – list to be sorted by `key` in ascending order
   - `desc` – list to be sorted by `key` in descending order
- sorts `list_of_person` by key in `order` using `method`.

Call the function `task2_5` with the contents of the file `PERSON.txt` and print the sorted `Person` objects using the following statements:

    list_of_person = task2_2('PERSON.txt')
    task2_5(list_of_person, 'quick sort', 'name', 'desc')
    for person in list_of_person:
        person.print() 
                                                                                                        [2]

In [24]:
def task2_5(list_of_person, method, key, order):
    if method == "insertion sort":
        task2_3(list_of_person, key, order)
    elif method == "quick sort":
        task2_4(list_of_person, key, order)
    else:
        print("Invalid Method")

list_of_person = task2_2('PERSON.txt')
task2_5(list_of_person, 'quick sort', 'name', 'desc')
for person in list_of_person:
    person.print()

(Name: Eva, age: 20)
(Name: Ethan, age: 17)
(Name: Emily, age: 19)
(Name: Ella, age: 18)
(Name: Eleanor, age: 16)
(Name: Dylan, age: 17)
(Name: Delilah, age: 20)
(Name: David, age: 16)
(Name: Daniel, age: 19)
(Name: Daisy, age: 18)
(Name: Cole, age: 20)
(Name: Chloe, age: 19)
(Name: Charlie, age: 17)
(Name: Catherine, age: 18)
(Name: Caleb, age: 16)
(Name: Brian, age: 19)
(Name: Bob, age: 20)
(Name: Benjamin, age: 16)
(Name: Bella, age: 17)
(Name: Bailey, age: 18)
(Name: Austin, age: 19)
(Name: Arthur, age: 20)
(Name: Amelia, age: 17)
(Name: Alice, age: 18)
(Name: Adam, age: 16)


### Task 3
A chess club wants to keep a record of players who registered for a team chess competition.
The record is implemented using Object-Oriented Programming (OOP).

#### task 3.1

The class `Player` is created with the following attributes:

- `name`, the name of the player
- `elo`, an integer representing the elo rating of the player
- `ptr` pointer, an integer pointing to the index of the next lower elo rating `Player` in the list


The class `PlayerList` contains three properties:

- `data` array of size `n`, with each element being a `Player` object
- `head` pointer, an integer pointing to the index of the first element in the linked list
- `free` pointer, an integer pointing to the index of the first element in the free list


The class `PlayerList` is created with the following methods:

- a constructor to set `head` to `-1`, `free` to `0`, and creates the `data` array with `n` empty `Player` nodes indicated with `name` set to '-' and `elo` set to `-1`
- `size()` method which returns the number of registered players
- `register(name, elo)` method which registers a player with `name` and `elo`, outputting a suitable error message when the `data` array is full
- `withdraw(name)` method which removes `name` from `PlayerList`, displaying an error message if name is not found
- `display()` method which displays the value of the `head` pointer, the value of the `free` pointer, and the contents of the `data` array in array index order.

Write the program code for the Player class and PlayerList class. [20]

In [43]:
class Player():
    def __init__(self, name="-", elo=-1, ptr=-1):
        self._name = name
        self._elo = elo
        self._ptr = ptr
    def __str__(self):
        return f"{self._name}: {self._elo} elo => {self._ptr}"
    def get_name(self):
        return self._name
    def get_elo(self):
        return self._elo
    def get_ptr(self):
        return self._ptr
    def set_name(self, name):
        self._name = name
    def set_elo(self, elo):
        self._elo = elo
    def set_ptr(self, ptr):
        self._ptr = ptr

class PlayerList():
    def __init__(self, n=10):
        data = [None] * n
        data[-1] = Player(ptr=-1)
        for i in reversed(range(n - 1)):
            player = Player(ptr=i+1)
            data[i] = player
        self._data = data
        self._head = -1
        self._free = 0

    def size(self):
        if self._head == -1:
            return 0
        probe = self._head
        size = 1
        while self._data[probe].get_ptr() != -1:
            size += 1
            probe = self._data[probe].get_ptr()
        return size

    def register(self, name, elo):
        if self._free == -1:
            print("Data array is full!")
        else:
            # Remove from free-space list
            new = self._free
            self._free = self._data[self._free].get_ptr()

            # Allocate new values
            self._data[new].set_name(name)
            self._data[new].set_elo(elo)

            # Prepend to head
            self._data[new].set_ptr(self._head)
            self._head = new

    def withdraw(self, name):
        previous, probe = -1, self._head
        while probe != -1 and self._data[probe].get_name() != name:
            previous, probe = probe, self._data[probe].get_ptr()
        if probe == -1:
            print(f"Name {name} not found")
        else:
            if previous == -1:
                # Target is head
                self._head = self._data[self._head].get_ptr()
            else:  
                # Cut out from head list
                nxt = self._data[probe].get_ptr()
                self._data[previous].set_ptr(nxt)
                
                # Add to free space list
                self._data[probe].set_ptr(self._free)
                self._free = probe

    def display(self):
        print(f"Head: {self._head}")
        print(f"Free: {self._free}")
        for i, player in enumerate(self._data):
            print(f"[{i}] {player}")

#### Task 3.2
The players who registered for the chess team have their name and elo rating recorded in a file named `CHESS.csv`.

Write program code to:

- create a `PlayerList` object, `cteam`, that accepts registration for up to 7 players
- read `CHESS.csv` and register them into `cteam`
- use `display()` to show the list of the players
- remove the player named `Taylor` from the team
- print the size of the team
- use `display()` to show the list of the players who are still in the team
                                                                                                [5]

In [50]:
cteam = PlayerList(7)
with open("CHESS.csv", "r") as file:
    # File closes automatically
    for line in file:
        line = line.strip()
        name, elo = tuple(line.split(","))
        cteam.register(name, elo)
cteam.display()

print()
cteam.withdraw("Taylor")
print(f"Size: {cteam.size()}")

print()
cteam.display()

Data array is full!
Data array is full!
Head: 6
Free: -1
[0] Nicki: 1250 elo => -1
[1] Lisa: 1337 elo => 0
[2] Iggy: 828 elo => 1
[3] Taylor: 1109 elo => 2
[4] Missy: 1437 elo => 3
[5] Megan: 745 elo => 4
[6] Cardi: 962 elo => 5

Size: 6

Head: 6
Free: 3
[0] Nicki: 1250 elo => -1
[1] Lisa: 1337 elo => 0
[2] Iggy: 828 elo => 1
[3] Taylor: 1109 elo => -1
[4] Missy: 1437 elo => 2
[5] Megan: 745 elo => 4
[6] Cardi: 962 elo => 5


### Task 4
A donut store owner currently keeps paper records about its members, donuts on sale and the purchase records by members. The store owner wants to create a suitable database to store the data and to allow them to run searches for specific data. The database will have three tables: a table to store data about the donuts, a table about the members and a table about the sales. The fields in each table are:

Donut:
- DonutID – donut’s unique number, for example, 5
- DonutName – donut’s name
- UnitPrice – price of one donut

Member:
- MemberNumber – member’s unique number, for example, 101
- MemberName – member’s name
- Phone – member’s contact number

Sale:
- SaleID – the purchase’s unique number, for example, 1030
- MemberNumber – the member’s unique number
- DonutID – the donut’s unique number
- Date – the date that the member purchased the donut, for example, '20230720'
- Quantity – the number of donuts purchased

#### Task 4.1
Write a Python program that uses SQL code to create the database `STORE` with the three tables given. Define the primary and foreign keys for each table. [4]

In [59]:
import sqlite3

con = sqlite3.connect("donut_store.db")
con.row_factory = sqlite3.Row
cur = con.cursor()

cur.execute("""
CREATE TABLE Donut(
    DonutID INTEGER PRIMARY KEY AUTOINCREMENT,
    DonutName TEXT,
    UnitPrice INTEGER
)
""")
cur.execute("""
CREATE TABLE Member(
    MemberNumber INTEGER PRIMARY KEY,
    MemberName TEXT,
    Phone INTEGER
)
""")
cur.execute("""
CREATE TABLE Sale(
    SaleID INTEGER PRIMARY KEY AUTOINCREMENT,
    MemberNumber INTEGER, 
    DonutID INTEGER, 
    Date TEXT,
    Quantity INTEGER,
    FOREIGN KEY(MemberNumber) REFERENCES Member(MemberNumber),
    FOREIGN KEY(DonutID) REFERENCES Donut(DonutID)
)
""")

cur.close()
con.close()

#### Task 4.2
The text files `DONUT.txt`, `MEMBER.txt`, and `SALE.txt` store the comma-separated values for each of the tables in the database.

Write a Python program to read in the data from each file and then store each item of data in the correct place in the database. [3]


In [77]:
import sqlite3

con = sqlite3.connect("donut_store.db")
con.row_factory = sqlite3.Row
cur = con.cursor()

with open("DONUT.txt", "r") as file:
    # File closes automatically
    for line in file:
        values = line.strip().split(",")        
        cur.execute("""
        INSERT INTO Donut(DonutID, DonutName, UnitPrice) VALUES (?, ?, ?)
        """, tuple(values))
con.commit()

with open("MEMBER.txt", "r") as file:
    # File closes automatically
    for line in file:
        values = line.strip().split(",")        
        cur.execute("""
        INSERT INTO Member(MemberNumber, MemberName, Phone) VALUES (?, ?, ?)
        """, tuple(values))
con.commit()

with open("SALE.txt", "r") as file:
    # File closes automatically
    for line in file:
        values = line.strip().split(",")        
        cur.execute("""
        INSERT INTO Sale(SaleID, MemberNumber, DonutID, Date, Quantity) VALUES (?, ?, ?, ?, ?)
        """, tuple(values))
con.commit()

cur.close()
con.close()

#### Task 4.3
Write a Python program to input a member’s number and display

- the member’s name,
- a table tabulating the donut names, dates and quantity of all the sales from this member

Test your program by running the application with the member number 104. [6]


In [86]:
import sqlite3

member_number = int(input("Enter Member Number: "))

con = sqlite3.connect("donut_store.db")
con.row_factory = sqlite3.Row
cur = con.cursor()

req = """
SELECT Donut.DonutName, Sale.Date, Sale.Quantity
FROM Sale
INNER JOIN Donut ON Sale.DonutID = Donut.DonutID
WHERE Sale.MemberNumber = ?
""".strip()

cur.execute(req, (member_number,))
results = cur.fetchall()

print()
print(f"Donut Name           Date      Quantity")
for row in results:
    print(f"{row['DonutName']: <20} {row['Date']}  {row['Quantity']}pcs")

Enter Member Number:  101



Donut Name           Date      Quantity
Sugar Cruller        20230720  2pcs
Coconut Fashion      20230720  1pcs
Ping Classic         20230721  2pcs
Ping Classic         20230721  3pcs
Banana Chocolate     20230723  3pcs


#### Task 4.4
The store owner wants to filter the purchases by Date and display the results in a web browser.

Write a Python program and the necessary files to create a web application that:

- receives a Date string from an HTML form,
- returns an HTML document that enables the web browser to display a table tabulating the names and the total quantity of each donut sold on that date, in descending order of the total quantity. 

Save your Python program as

`Task4_4_<your name>_<centre number>_<index number>.py`

with any additional files / sub-folders as needed in a folder named

`Task4_4_<your name>_<centre number>_<index number>`

Run the web application with the date entered as '20230721'. Save the output as

`Task4_4_<your name>_<centre number>_<index number>.html`
                                                                                                [12]